src/Features/CSharp/Portable/Diagnostics/Analyzers/TypeSyntaxSimplifierWalker_Binding.cs

```diff
@@ -0,0 +1,354 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.SimplifyTypeNames
+{
+    /// <summary>
+    /// This walker sees if we can simplify types/namespaces that it encounters. Importantly, it
+    /// only checks types/namespaces in contexts that are known to only allows types/namespaces only
+    /// (i.e. declarations, casts, etc.).  It does not check general expression contexts.
+    /// <para/>
+    /// A core concept here is that we'd like to perform as little binding as possible. So this
+    /// walker builds up information as it walks the tree (like what names are in scope) so it can
+    /// avoid binding nodes it knows cannot be simplified at all.
+    /// </summary>
+    internal partial class TypeSyntaxSimplifierWalker : CSharpSyntaxWalker
```

Comment:
the general idea here is that we can keep track of what is in scope and then try to syntactically bail from more complex analysis as much as possible.